app/main/views/frameworks.py

```diff
@@ -566,3 +595,219 @@ def upload_framework_agreement(framework_slug):
+                "signaturepage.fail: signerName:{signerName} signerRole:{signerRole} {errors}",
+                extra={
+                    'signerName': form['signerName'].data,
+                    'signerRole': form['signerRole'].data,
+                    'errors': ",".join(chain.from_iterable(form.errors.values()))
+                }
+            )
+
+            error_keys_in_order = [key for key in question_keys if key in form.errors.keys()]
+            form_errors = [
+                {'question': form[key].label.text, 'input_name': key} for key in error_keys_in_order
+            ]
+
+    # if the signer* keys exist, prefill them in the form
+    if supplier_framework['agreementDetails']:
+        for question_key in question_keys:
+            if question_key in supplier_framework['agreementDetails']:
+                form[question_key].data = supplier_framework['agreementDetails'][question_key]
+
+    return render_template(
+        "frameworks/signer_details.html",
+        form=form,
+        form_errors=form_errors,
+        framework=framework,
+        question_keys=question_keys,
+        supplier_framework=supplier_framework,
+    ), 400 if form_errors else 200
+
+
+@main.route('/frameworks/<framework_slug>/signature-upload', methods=['GET', 'POST'])
+@login_required
+def signature_upload(framework_slug):
+    framework = get_framework(data_api_client, framework_slug)
+    return_supplier_framework_info_if_on_framework_or_abort(data_api_client, framework_slug)
+    agreements_bucket = s3.S3(current_app.config['DM_AGREEMENTS_BUCKET'])
+    signature_page = get_most_recently_uploaded_agreement_file_or_none(agreements_bucket, framework_slug)
+    upload_error = None
+
+    if request.method == 'POST':
+        # No file chosen for upload and file already exists on s3 so can use existing and progress
+        if not request.files['signature_page'].filename and signature_page:
+            return redirect(url_for(".contract_review", framework_slug=framework_slug))
+
+        if not file_is_image(request.files['signature_page']) and not file_is_pdf(request.files['signature_page']):
+            upload_error = "The file must be a PDF, JPG or PNG"
+        elif not file_is_less_than_5mb(request.files['signature_page']):
+            upload_error = "The file must be less than 5MB"
+        elif file_is_empty(request.files['signature_page']):
+            upload_error = "The file must not be empty"
+
+        if not upload_error:
+            upload_path = get_agreement_document_path(
+                framework_slug,
+                current_user.supplier_id,
+                '{}{}'.format(SIGNED_AGREEMENT_PREFIX, get_extension(request.files['signature_page'].filename))
+            )
+            agreements_bucket.save(
+                upload_path,
+                request.files['signature_page'],
+                acl='private'
+            )
+
+            session['signature_page'] = request.files['signature_page'].filename
```

Comment:
Decision reversed after watching user testing (aware Cath said otherwise in show and tell yesterday but she had not been updated on that decision when presenting)